File: Step 06 LinkedList/Step 6.2 Doubly Linked List/Doubly linked list Insertion at given position/Doubly linked list Insertion at given position.java


/* Structure of Doubly Linked List
class Node
{
	int data;
	Node next;
	Node prev;
	Node(int data)
	{
	    this.data = data;
	    next = prev = null;
	}
}*/

class GfG
{
    //Function to insert a new node at given position in doubly linked list.
    void addNode(Node head_ref, int pos, int data)
	{
		// Your code here
		Node ptr = new Node(data);
		Node temp = head_ref;
		int i = 0;
		while( i<pos ){
		    temp = temp.next;
		    i++;
		}
		ptr.next = temp.next;
		if(temp.next!=null) temp.next.prev = ptr;
		ptr.prev = temp;
		temp.next = ptr;
	}
}

File: Step 06 LinkedList/Step 6.2 Doubly Linked List/Doubly linked list Insertion at given position/Doubly linked list Insertion at given position.cpp


/* a Node of the doubly linked list 
struct Node
{
  int data;
  struct Node *next;
  struct Node *prev;
  Node(int x) { data = x; next = prev = NULL; }
}; */

//Function to insert a new node at given position in doubly linked list.
void addNode(Node *head, int pos, int data)
{
   // Your code here
    if(head == NULL) return ;
   int count=-1;
   Node * temp= head;
   while(temp->next!=NULL){
       count++;
       if(count ==pos)break;
       temp=temp->next;
       
   }
   if(temp->next ==NULL) ++count;
   Node * newnode =new Node(data);
   Node *nextnode=temp->next;
   if(nextnode == NULL && count==pos){//insertion after tail
       newnode->prev=temp;
       temp->next =newnode;
       return ;
   }
   
   temp->next=newnode;
   newnode->prev=temp;
   nextnode->prev=newnode;
   newnode->next=nextnode;
   return ;
}